### ISSUE TITLE
Run slang-test on an Intel GPU and file slang issues

### ISSUE THREAD

**Issue by cheneym2** [shader-slang/slang#5601]
If there are any Slang issues running slang-test on Intel GPUs, they need to be reported so we can improve Slang on that platform.

**Comment by aleino-nv** [comment]
Lots of tests are failing when I do a full slang-test run locally, without switching to the Intel GPU.
I'll look into those failures first.

**Comment by aleino-nv** [comment]
Some results from my dev machine, which has an integrated Intel GPU and a discrete NVIDIA GPU.

Commit 8000e0ede34e920cc7f37d69a335d74a472eff42.

NVIDIA: NVIDIA RTX A1000 (Driver version 552.41)
Intel: Intel Iris Xe (Driver version 31.0.101.5333)

All testing is done with Release build.
Switching between NVIDIA and Intel is done by setting per-app GPU preference for both slang-test.exe and test-server.exe via `System > Display > Graphics`.

Failures: 
- `slang-unit-test-tool/comHostCallable.internal`: Intel, NVIDIA
- `tests/bugs/dxbc-double-problem.slang (dx11)`: Intel
- `tests/bugs/frexp-double.slang (dx11)`: Intel
- `tests/compute/atomics-buffer.slang (dx11)`: Intel
- `tests/compute/texture-get-dimensions.slang.1 (dx11)`: Intel
- `tests/compute/transcendental-double.slang.2 (dx11)`: Intel
- `tests/diagnostics/local-line.slang.3`: Intel, NVIDIA
- `tests/diagnostics/syntax-error-intrinsic.slang.3`: Intel, NVIDIA
- `tests/diagnostics/syntax-error-op-line-2.slang.3`: Intel, NVIDIA
- `tests/diagnostics/syntax-error-op-line-3.slang.3`: Intel, NVIDIA
- `tests/diagnostics/syntax-error-op-line.slang.3`: Intel, NVIDIA
- `tests/hlsl-intrinsic/bit-cast-double.slang.1 (dx11)`: Intel
- `tests/hlsl-intrinsic/classify-double.slang.1 (dx11)`: Intel
- `tests/hlsl-intrinsic/classify-float.slang.1 (dx11)`: Intel
- `tests/hlsl-intrinsic/scalar-double-simple.slang.1 (dx11)`: Intel

**Comment by aleino-nv** [comment]
For `tests/bugs/dxbc-double-problem.slang` [ID3D11Device::CreateComputeShader](https://learn.microsoft.com/en-us/windows/win32/api/d3d11/nf-d3d11-id3d11device-createcomputeshader) is returning E_INVALIDARG on Intel. I guess it doesn't like the DXBC code we pass it.

**Comment by aleino-nv** [comment]
For `tests/compute/atomics-buffer.slang (dx11)` the output is wrong.
I'm leaning towards 'Intel driver bug' on this one.


    EXPECTED{{{
    210
    301
    32
    123
    }}}
    ACTUAL{{{
    666
    0
    0
    0
    0
    0
    0
    0
    0
    0
    0
    0
    0
    0
    0
    0
    }}}


Here is the HLSL:

    #pragma pack_matrix(column_major)
    #ifdef SLANG_HLSL_ENABLE_NVAPI
    #include "nvHLSLExtns.h"
    #endif
    
    #ifndef __DXC_VERSION_MAJOR
    // warning X3557: loop doesn't seem to do anything, forcing loop to unroll
    #pragma warning(disable : 3557)
    #endif
    
    
    #line 22 "tests/compute/atomics-buffer.slang"
    RWBuffer<uint > outputBuffer_0 : register(u0);
    
    void test_0(uint val_0)
    {
    
    
        uint _S1;
        InterlockedAdd(outputBuffer_0[val_0], val_0, _S1);
    
    #line 29
        uint _S2 = val_0 * 16U;
    
    #line 29
        uint _S3;
        InterlockedAdd(outputBuffer_0[val_0 ^ 1U], _S2, _S3);
    
    #line 30
        uint _S4;
        InterlockedAdd(outputBuffer_0[val_0 ^ 2U], _S2 * 16U, _S4);
    
    #line 31
        return;
    }
    
    [numthreads(4, 1, 1)]
    void computeMain(uint3 dispatchThreadID_0 : SV_DispatchThreadID)
    {
        test_0(dispatchThreadID_0.x);
        return;
    }

**Comment by aleino-nv** [comment]
For `tests/hlsl-intrinsic/classify-float.slang.1 (dx11)` the output at (zero based) index `3` is `0` (meaning `isnan, isfinite, isinf = false, false, false`) when `1` (meaning `isnan, isfinite, isinf = true, false, false`) is expected for the input `0x7fffffff`.

The output `HLSLS` is calling these exact `HLSL` functions, so I'm leaning toward 'Intel driver bug' on this one.

**Comment by aleino-nv** [comment]
@cheneym2 @csyonghe Out of the above Intel-specific findings, only the one about the [rejected DXBC code](https://github.com/shader-slang/slang/issues/5601#issuecomment-2609240712) for those 'double' tests is something that I'd look a bit closer at, from a Slang quality perspective. 

Other than that, one mystery for me is: [why 8 MIP levels and not 4](https://github.com/shader-slang/slang/issues/5601#issuecomment-2609485992)?

**Comment by aleino-nv** [comment]
Yeah [this issue](https://github.com/shader-slang/slang/issues/5601#issuecomment-2609240712) seems pretty clearly to be an Intel driver problem:

    RWStructuredBuffer<double> outputBuffer;

    [numthreads(1, 1, 1)]
    void main(uint3 dispatchThreadID : SV_DispatchThreadID)
    {
        outputBuffer[0] = double(dispatchThreadID.x);
    }

The above shader is accepted when I run on my NVIDIA GPU and rejected with `E_INVALIDARG` on Intel.

**Comment by cheneym2** [comment]
Is the double precision issue same as https://community.intel.com/t5/GPU-Compute-Software/Iris-Xe-Graphics-can-t-compute-double-precision-floating-point/m-p/1281976 ?

**Comment by aleino-nv** [comment]
> Is the double precision issue same as https://community.intel.com/t5/GPU-Compute-Software/Iris-Xe-Graphics-can-t-compute-double-precision-floating-point/m-p/1281976 ?

Yeah, I think probably this is the same underlying issue.

According to footnote 2 [on this page](https://learn.microsoft.com/en-us/windows/win32/direct3d11/overviews-direct3d-11-devices-downlevel-intro), 

> Shader model 5.0 and above can optionally support double-precision shaders, extended double-precision shaders, the SAD4 shader instruction, and partial-precision shaders. To determine the shader model 5.0 options that are available for DirectX 11, call [ID3D11Device::CheckFeatureSupport](https://learn.microsoft.com/en-us/windows/win32/api/D3D11/nf-d3d11-id3d11device-checkfeaturesupport). Some compatibility depends on what hardware you are running on. Shader model 5.1 and above are only supported through the DirectX 12 API, regardless of the feature level that's being used. DirectX 11 only supports up to shader model 5.0. The DirectX 12 API only goes down to feature level 11_0.

When I query D3D11_FEATURE_DOUBLES on my Intel GPU, it indeed reports that doubles are not supported.
We should probably query for this when determining if the test is supported or not, instead of reporting a failure.

**Comment by aleino-nv** [comment]
I filed https://github.com/shader-slang/slang/issues/6171 for the doubles-issue.

**Comment by aleino-nv** [comment]
> [#5601 (comment)](https://github.com/shader-slang/slang/issues/5601#issuecomment-2609485992) might be worth looking closer at. Why 8?

Yeah, generally I don't really understand the `mipCount` output of [tests/compute/texture-get-dimensions.slang](https://github.com/shader-slang/slang/blob/master/tests/compute/texture-get-dimensions.slang).

According to the [expected output file](https://github.com/shader-slang/slang/blob/master/tests/compute/texture-get-dimensions.slang.expected.txt):

    //TEST_INPUT: Texture1D(size=4, content = one):name t1D
    Texture1D<float> t1D;                                                             // mipCount = 0 ???     (I would expect log2(4)+1 = 3)
    //TEST_INPUT: Texture2D(size=8, content = one):name t2D
    Texture2D<float> t2D;                                                             // mipCount = 0 ???     (I would expect log2(8)+1 = 4)
    //TEST_INPUT: Texture3D(size=2, content = one):name t3D
    Texture3D<float> t3D;                                                             // mipCount = 0x80 ???  (I would expect log2(2)+1 = 2)
    //TEST_INPUT: TextureCube(size=16, content = one):name tCube
    TextureCube<float> tCube;                                                         // mipCount = 0 ???     (I would expect log2(16)+1 = 5)
    //TEST_INPUT: Texture1D(size=2, content = one, arrayLength=2):name t1DArray
    Texture1DArray<float> t1DArray;                                                   // mipCount = 0x80 ???  (I would expect log2(2)+1 = 2)
    //TEST_INPUT: Texture2D(size=4, content = one, arrayLength=3):name t2DArray
    Texture2DArray<float> t2DArray;                                                   // mipCount = 0x60 ???  (I would expect log2(4)+1 = 3)
    //TEST_INPUT: TextureCube(size=16, content = one, arrayLength=1):name tCubeArray
    TextureCubeArray<float> tCubeArray;                                               // mipCount = 0x08 ???  (I would expect log2(16)+1 = 5)

**Comment by aleino-nv** [comment]
In any case, in terms of Intel specific Slang issues I conclude that I was not able to find any major ones.
The https://github.com/shader-slang/slang/issues/6171 issue is relatively minor because it's mostly about `slang-test` itself and not really about Slang's output.

Closing the task.

**Comment by jkwak-work** [comment]
I think the mipcount might be a bug in slang-rhi.
